funsor/pyro/hmm.py

```diff
@@ -156,3 +167,114 @@ def expand(self, batch_shape, _instance=None):
         new._obs = self._obs
         super(GaussianMRF, new).__init__(self.funsor_dist, batch_shape, self.event_shape)
         return new
+
+
+class GaussianDiscreteMRF(FunsorDistribution):
+    """
+    Temporal Markov Random Field with Gaussian latent state and discrete
+    observations, where observation factors are mixtures-of-gaussians.
+
+    This distribution uses the :func:`~funsor.terms.moment_matching`
+    approximation to perform inference in the following Pyro model, where
+    ``state`` be the latent multivariate normal state, ``prediction`` is the
+    predicted future state, and ``obs`` is the discrete observation::
+
+        def model(data):
+            state = pyro.sample("state_0", initial_dist)
+            for t in range(len(data)):
+                noise = pyro.sample("noise_{}".format(t), transition_dist)
+                state = state @ transition_matrix[t] + noise
+                obs = pyro.sample("x_{}".format(t),
+                                  dist.Categorical(observation_logits),
+                                  obs=data[t])
+                pyro.sample("state_{}".format(t),
+                            observation_dist[t, data[t]],
+                            obs=state)
+
+    :param ~torch.distributions.MultivariateNormal initial_dist: Represents
+        ``p(state[0])``.
+    :param ~torch.Tensor transition_matrix: Transforms ``state[t]`` to
+        ``prediction[t+1]``.
+    :param ~torch.distributions.MultivariateNormal transition_dist: Represents
+        ``p(state[t+1] | prediction[t+1])``.
+    :param ~torch.Tensor observation_logits: Represents ``p(obs[t+1])``.
+    :param ~torch.distributions.MultivariateNormal observation_dist: Represents
+        ``p(state[t+1] | obs[t+1])``.
```

Comment:
@fritzo I am trying to understand this model. Could you confirm that the docs above is correct?

Reply:
The docstrings are correct (except for minor off-by-1 time shifts; I'll fix those shortly).

I agree this is kind of a weird model: it is not directed and not normalized. However as a CRF/MRF it is a well-posed non-normalized density and admits tractable approximate inference via moment matching. To normalize we use the same trick as with the `GaussianMRF` in Pyro.

Reply:
I'm also open to other models that admit approximate inference, in case you can think of any. Eventually we'd like to implement a `linearize` interpretation #89 which would enable Extended Kalman Filter -like inference, e.g. linear-gaussian hidden dynamics and a probit or logistic observation model. I've started with the CRF/MRF in this PR because we already have already implemented the `moment_matching` interpretation #115 .

Reply:
Thanks! I think I start understanding now. :D